File: kt/src/main/kotlin/com/github/ephemient/aoc2025/Day3.kt
package com.github.ephemient.aoc2025

object Day3 {
    fun day3(input: String, digits: Int): Long = input.lines().sumOf { line ->
        if (line.length < digits) return@sumOf 0
        var startIndex = 0
        var acc = 0L
        for (i in 1..digits) {
            val c = line.substring(startIndex, line.length - digits + i).max()
            startIndex = line.indexOf(c, startIndex) + 1
            acc = 10 * acc + c.digitToInt()
        }
        acc
    }

    fun part1(input: String): Long = day3(input, 2)
    fun part2(input: String): Long = day3(input, 12)
}

File: rs/src/day3.rs
use std::cmp::Reverse;

fn day3(input: &str, digits: usize) -> usize {
    input
        .lines()
        .filter_map(|line| {
            if line.len() < digits {
                return None;
            }
            let line = line.as_bytes();
            let mut start = 0;
            let mut acc = 0;
            for n in (0..digits).rev() {
                let (i, b) = line[start..line.len() - n]
                    .iter()
                    .enumerate()
                    .min_by_key(|(_, b)| Reverse(**b))?;
                start += i + 1;
                acc = 10 * acc + (*b as char).to_digit(10)? as usize;
            }
            Some(acc)
        })
        .sum::<usize>()
}

pub fn part1(input: &str) -> usize {
    day3(input, 2)
}

pub fn part2(input: &str) -> usize {
    day3(input, 12)
}
